### ISSUE TITLE
bug(hubble): DB can be corrupted if connection is reset while downloading snapshot

### ISSUE THREAD

**Issue by sanjayprabhu** [farcasterxyz/hub-monorepo#2042]
**What is the bug?**
If the connection is interrupted when downloading the snapshot, the db is corrupted and hubble doesn't recover. We should either retry on connection reset, or detect this case and recover gracefully.

```
hubble-1  | {"message":"{\"level\":30,\"time\":1718039666691,\"pid\":29,\"hostname\":\"a0a85eb82e41\",\"chunkCount\":25,\"totalChunks\":32,\"msg\":\"Downloading snapshot chunks...\"}\n","timestamp":"2024-06-10T17:14:26.691Z","type":"out","process_id":0,"app_name":"hubble"}
hubble-1  | {"message":"{\"level\":30,\"time\":1718039666691,\"pid\":29,\"hostname\":\"a0a85eb82e41\",\"chunkCount\":26,\"totalChunks\":32,\"msg\":\"Downloading snapshot chunks...\"}\n","timestamp":"2024-06-10T17:14:26.691Z","type":"out","process_id":0,"app_name":"hubble"}
hubble-1  | {"message":"{\"level\":50,\"time\":1718039666691,\"pid\":29,\"hostname\":\"a0a85eb82e41\",\"error\":{\"code\":\"ECONNRESET\"},\"msg\":\"Error extracting snapshot\"}\n","timestamp":"2024-06-10T17:14:26.691Z","type":"out","process_id":0,"app_name":"hubble"}
hubble-1  | {"message":"{\"level\":50,\"time\":1718039666691,\"pid\":29,\"hostname\":\"a0a85eb82e41\",\"error\":{\"errCode\":\"unavailable\",\"presentable\":false,\"name\":\"HubError\"},\"msg\":\"failed to sync snapshot, falling back to regular sync\"}\n","timestamp":"2024-06-10T17:14:26.691Z","type":"out","process_id":0,"app_name":"hubble"}
hubble-1  | {"message":"{\"level\":30,\"time\":1718039666691,\"pid\":29,\"hostname\":\"a0a85eb82e41\",\"msg\":\"attempting catchup sync with snapshot\"}\n","timestamp":"2024-06-10T17:14:26.691Z","type":"out","process_id":0,"app_name":"hubble"}
hubble-1  | {"message":"{\"level\":30,\"time\":1718039666691,\"pid\":29,\"hostname\":\"a0a85eb82e41\",\"currentItemCount\":415158543,\"delta\":143933,\"limit\":25000000,\"msg\":\"catchup sync using snapshot not required\"}\n","timestamp":"2024-06-10T17:14:26.691Z","type":"out","process_id":0,"app_name":"hubble"}
hubble-1  | {"message":"{\"level\":50,\"time\":1718039666691,\"pid\":29,\"hostname\":\"a0a85eb82e41\",\"msg\":\"catchup sync skipped\"}\n","timestamp":"2024-06-10T17:14:26.691Z","type":"out","process_id":0,"app_name":"hubble"}
hubble-1  | {"message":"{\"level\":50,\"time\":1718039666691,\"pid\":29,\"hostname\":\"a0a85eb82e41\",\"retryCount\":1,\"error\":{},\"errorMessage\":\"db.internal_error/Corruption: Corruption: IO error: No such file or directory: While open a file for random read: .rocks/rocks.hub._default/004268.sst: No such file or directory  The file .rocks/rocks.hub._default/MANIFEST-000005 may be corrupted.\",\"msg\":\"failed to open rocksdb. Retry in 15s\"}\n","timestamp":"2024-06-10T17:14:26.691Z","type":"out","process_id":0,"app_name":"hubble"}
hubble-1  | {"message":"{\"level\":50,\"time\":1718039666691,\"pid\":29,\"hostname\":\"a0a85eb82e41\",\"retryCount\":2,\"error\":{},\"errorMessage\":\"db is opening\",\"msg\":\"failed to open rocksdb. Retry in 15s\"}\n","timestamp":"2024-06-10T17:14:26.691Z","type":"out","process_id":0,"app_name":"hubble"}
hubble-1  | {"message":"{\"level\":50,\"time\":1718039666691,\"pid\":29,\"hostname\":\"a0a85eb82e41\",\"retryCount\":3,\"error\":{},\"errorMessage\":\"db is opening\",\"msg\":\"failed to open rocksdb. Retry in 15s\"}\n","timestamp":"2024-06-10T17:14:26.691Z","type":"out","process_id":0,"app_name":"hubble"}
hubble-1  | {"message":"{\"level\":50,\"time\":1718039666691,\"pid\":29,\"hostname\":\"a0a85eb82e41\",\"retryCount\":4,\"error\":{},\"errorMessage\":\"db is opening\",\"msg\":\"failed to open rocksdb. Retry in 15s\"}\n","timestamp":"2024-06-10T17:14:26.691Z","type":"out","process_id":0,"app_name":"hubble"}
hubble-1  | {"message":"{\"level\":50,\"time\":1718039666691,\"pid\":29,\"hostname\":\"a0a85eb82e41\",\"retryCount\":5,\"error\":{},\"errorMessage\":\"db is opening\",\"msg\":\"failed to open rocksdb. Retry in 15s\"}\n","timestamp":"2024-06-10T17:14:26.691Z","type":"out","process_id":0,"app_name":"hubble"}
hubble-1  | {"message":"{\"level\":30,\"time\":1718039666691,\"pid\":29,\"hostname\":\"a0a85eb82e41\",\"msg\":\"Stopping Hubble...\"}\n","timestamp":"2024-06-10T17:14:26.691Z","type":"out","process_id":0,"app_name":"hubble"}

```

**How can it be reproduced? (optional)**
Interrupt the connection when snapshot is being downloaded

**Additional context (optional)**

**Comment by suraneti** [comment]
I also faced this issue, db was corrupted if the connection broke.

**Comment by maxpetretta** [comment]
ran into this as well, some more details here: https://warpcast.com/maxp.eth/0x67b6d676

**Comment by JayJiangEth** [comment]
See the same issue, my computer close by mistake, how can I restart this?

**Comment by vvzvlad** [comment]
I have the same behavior. 
```
 Starting Hubble
------------------
Please wait... This may take several minutes
{"level":60,"time":1721136330997,"pid":30,"hostname":"4fc841bda700","err":{"type":"Error","message":"Failed to start hub: db.internal_error/Corruption: Corruption: IO error: No such file or directory: While open a file for random read: .rocks/rocks.hub._default/trieDb/001412.sst: No such file or directory  The file .rocks/rocks.hub._default/trieDb/MANIFEST-000005 may be corrupted.","stack":"Error: Failed to start hub\n    at file:///home/node/app/apps/hubble/src/cli.ts:701:14\n    at /home/node/app/node_modules/neverthrow/dist/index.cjs.js:113:102\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at Command.<anonymous> (file:///home/node/app/apps/hubble/src/cli.ts:699:25)\ncaused by: Error: db.internal_error/Corruption: Corruption: IO error: No such file or directory: While open a file for random read: .rocks/rocks.hub._default/trieDb/001412.sst: No such file or directory  The file .rocks/rocks.hub._default/trieDb/MANIFEST-000005 may be corrupted.\n    at rsDbOpen (file:///home/node/app/apps/hubble/src/rustfunctions.ts:112:14)\n    at file:///home/node/app/apps/hubble/src/storage/db/rocksdb.ts:160:9\n    at new Promise (<anonymous>)\n    at RocksDB.open (file:///home/node/app/apps/hubble/src/storage/db/rocksdb.ts:151:12)\n    at Function.numItems (file:///home/node/app/apps/hubble/src/network/sync/merkleTrie.ts:129:14)\n    at Hub.attemptCatchupSyncWithSnapshot (file:///home/node/app/apps/hubble/src/hubble.ts:904:40)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at Hub.start (file:///home/node/app/apps/hubble/src/hubble.ts:601:27)\n    at Command.<anonymous> (file:///home/node/app/apps/hubble/src/cli.ts:699:25)"},"msg":"Failed to start hub"}
{"level":60,"time":1721136331129,"pid":30,"hostname":"4fc841bda700","reason":"Hub Startup failed","msg":"shutting down hub"}
{"level":30,"time":1721136332430,"pid":30,"hostname":"4fc841bda700","server":{"address":"statsd","port":8125,"family":4},"msg":"Statsd server specified. Statsd enabled"}
{"level":30,"time":1721136332430,"pid":30,"hostname":"4fc841bda700","msg":"StorageRegistry: 0x00000000fcce7f938e7ae6d3c335bd6a1a7c593d, KeyRegistry: 0x00000000Fc1237824fb747aBDE0FF18990E59b7e, IdRegistry: 0x00000000Fc6c5F01Fc30151999387Bb99A9f489b"}
{"level":30,"time":1721136332430,"pid":30,"hostname":"4fc841bda700","totalPruneSize":10080,"msg":"total default storage limit size"}
{"level":30,"time":1721136332430,"pid":30,"hostname":"4fc841bda700","rpcRateLimit":20000,"msg":"RPC rate limit enabled"}
{"level":30,"time":1721136332430,"pid":30,"hostname":"4fc841bda700","component":"utils/p2p","ip":"46.188.5.21","msg":"Fetched public IP"}
{"level":30,"time":1721136332431,"pid":30,"hostname":"4fc841bda700","msg":"attempting catchup sync with snapshot"}
{"level":30,"time":1721136332431,"pid":30,"hostname":"4fc841bda700","msg":"Stopping Hubble..."}
{"level":30,"time":1721136332431,"pid":30,"hostname":"4fc841bda700","msg":"Stopped http API server"}
{"level":30,"time":1721136332431,"pid":30,"hostname":"4fc841bda700","msg":"gRPC server force shutdown succeeded"}
{"level":30,"time":1721136332431,"pid":30,"hostname":"4fc841bda700","msg":"Stopped watching contract events"}
{"level":30,"time":1721136332431,"pid":30,"hostname":"4fc841bda700","msg":"Stopped watching contract events"}
{"level":30,"time":1721136332431,"pid":30,"hostname":"4fc841bda700","msg":"Stopped watching contract events"}
{"level":30,"time":1721136332431,"pid":30,"hostname":"4fc841bda700","msg":"Stopped watching block numbers"}
{"level":30,"time":1721136332431,"pid":30,"hostname":"4fc841bda700","msg":"Stopped fname events provider"}
{"level":30,"time":1721136332431,"pid":30,"hostname":"4fc841bda700","msg":"L2EventsProvider stopped"}
{"level":30,"time":1721136332431,"pid":30,"hostname":"4fc841bda700","msg":"Sync engine stopped"}
{"level":30,"time":1721136332431,"pid":30,"hostname":"4fc841bda700","msg":"Stopped libp2p..."}
{"level":30,"time":1721136332431,"pid":30,"hostname":"4fc841bda700","msg":"stopping engine"}
{"level":30,"time":1721136332431,"pid":30,"hostname":"4fc841bda700","msg":"engine stopped"}
{"level":30,"time":1721136332431,"pid":30,"hostname":"4fc841bda700","msg":"Hubble stopped, exiting normally"}
```

**Comment by sanjayprabhu** [comment]
This should be fixed by https://github.com/farcasterxyz/hub-monorepo/pull/2236, included in the latest release (1.14.3).
